src/hardware/dac.rs

```diff
@@ -0,0 +1,279 @@
+    pub sync0: PG3<Output<PushPull>>,
+    pub sync1: PG2<Output<PushPull>>,
+    pub sync2: PG1<Output<PushPull>>,
+    pub sync3: PG0<Output<PushPull>>,
+    pub shdn0: PG4<Output<PushPull>>,
+    pub shdn1: PG5<Output<PushPull>>,
+    pub shdn2: PG6<Output<PushPull>>,
+    pub shdn3: PG7<Output<PushPull>>,
+}
+
+/// DAC: https://www.analog.com/media/en/technical-documentation/data-sheets/AD5680.pdf
+/// Peltier Driver: https://datasheets.maximintegrated.com/en/ds/MAX1968-MAX1969.pdf
+pub struct Dac {
+    spi: Spi<SPI3, Enabled, u8>,
+    pub val: [u32; 4],
+    gpio: DacGpio,
+}
+
+impl Dac {
+    pub fn new(
+        clocks: &CoreClocks,
+        prec: rec::Spi3,
+        spi3: SPI3,
+        sck: PC10<Alternate<AF6>>,
+        mosi: PC12<Alternate<AF6>>,
+        gpio: DacGpio,
+    ) -> Self {
+        let spi = spi3.spi(
+            (sck, NoMiso, mosi),
+            MODE_1,
+            SPI_CLOCK,
+            prec,
+            clocks, // default pll1_q clock source
+        );
+
+        let mut dac = Dac {
+            spi,
+            val: [0, 0, 0, 0],
+            gpio,
+        };
+
+        dac.gpio.sync0.set_high().unwrap();
+        dac.gpio.sync1.set_high().unwrap();
+        dac.gpio.sync2.set_high().unwrap();
+        dac.gpio.sync3.set_high().unwrap();
+
+        // default to zero amps
+        dac.set(0.0, Channel::Ch0);
+        dac.set(0.0, Channel::Ch1);
+        dac.set(0.0, Channel::Ch2);
+        dac.set(0.0, Channel::Ch3);
+        dac
+    }
+
+    /// Set the DAC output to current on a channel.
+    pub fn set(&mut self, curr: f32, ch: Channel) {
+        let value = i_to_dac(curr);
+        let value = value.min(MAX_VALUE);
+        // 24 bit transfer. First 4 bit and last 2 bit are low.
+        let mut buf = [(value >> 14) as u8, (value >> 6) as u8, (value << 2) as u8];
+
+        match ch {
+            Channel::Ch0 => {
+                self.gpio.sync0.set_low().unwrap();
+                self.spi.transfer(&mut buf).unwrap();
+                self.val[0] = value;
+                self.gpio.sync0.set_high().unwrap();
+            }
+            Channel::Ch1 => {
+                self.gpio.sync1.set_low().unwrap();
+                self.spi.transfer(&mut buf).unwrap();
+                self.val[1] = value;
+                self.gpio.sync1.set_high().unwrap();
+            }
+            Channel::Ch2 => {
+                self.gpio.sync2.set_low().unwrap();
+                self.spi.transfer(&mut buf).unwrap();
+                self.val[2] = value;
+                self.gpio.sync2.set_high().unwrap();
+            }
+            Channel::Ch3 => {
+                self.gpio.sync3.set_low().unwrap();
+                self.spi.transfer(&mut buf).unwrap();
+                self.val[3] = value;
+                self.gpio.sync3.set_high().unwrap();
+            }
+        }
+    }
+
+    /// enable or disable a TEC channel via shutdown pin.
+    pub fn en_dis_ch(&mut self, ch: Channel, en: bool) {
```

Comment:
`en_dis_ch` is three abbreviations. Please find a better name. `set_shutdown(&mut self, channel: Channel, shutdown: bool)` would be consistently traceable from datasheet through schematic through this code all they way into the mqtt namespace.